File: src/com/qubling/sidekick/util/HttpUtils.java
package com.qubling.sidekick.util;

import java.io.IOException;
import java.io.InputStreamReader;

import org.apache.http.HttpEntity;
import org.apache.http.HttpResponse;

public class HttpUtils {
    private HttpUtils() {}

    public static String getCharset(HttpResponse res) {
        HttpEntity entity = res.getEntity();

        String charset;
        String contentType = entity.getContentType().getValue();
        int charsetIndex = contentType.indexOf("charset=");
        if (charsetIndex >= 0) {
            int endingIndex = contentType.indexOf(";", charsetIndex + 8);
            if (endingIndex >= 0) {
                charset = contentType.substring(charsetIndex + 8, endingIndex);
            }
            else {
                charset = contentType.substring(charsetIndex + 8);
            }
        }
        else {
            charset = "UTF-8";
        }

        return charset;
    }
  
    public static String slurpContent(HttpResponse res) throws IOException {
        HttpEntity entity = res.getEntity();
        String charset = getCharset(res);

        InputStreamReader contentReader = new InputStreamReader(entity.getContent(), charset);
        char[] buf = new char[1000];
        StringBuilder contentStr = new StringBuilder();
        int readLength;
        while ((readLength = contentReader.read(buf)) > -1) {
            contentStr.append(buf, 0, readLength);
        }

        return contentStr.toString();
    }
}

File: src/main/scala/com/qubling/sidekick/util/HttpUtils.scala
package com.qubling.sidekick.util

import java.io.IOException
import java.io.InputStreamReader

import org.apache.http.HttpEntity
import org.apache.http.HttpResponse

import android.util.Log

import scala.io.Source

object HttpUtils {
  def getCharset(res : HttpResponse) : String = {
    val entity = res.getEntity

    val contentType = entity.getContentType.getValue
    val charsetIndex = contentType.indexOf("charset=")
    if (charsetIndex >= 0) {
      val endingIndex = contentType.indexOf(";", charsetIndex + 8)
      if (endingIndex >= 0) {
        Log.d("HttpUtils", "charset partial")
        contentType.substring(charsetIndex + 8, endingIndex)
      }
      else {
        Log.d("HttpUtils", "charset whole")
        contentType.substring(charsetIndex + 8)
      }
    }
    else {
      Log.d("HttpUtils", "fallback to UTF-8")
      "UTF-8"
    }
  }
  
  def slurpContent(res : HttpResponse) : String = {
    val entity = res.getEntity
    val charset = getCharset(res)

    Log.d("HttpUtils", "charset = " + charset)

    Source.fromInputStream(entity.getContent, charset).mkString("")
  }
}
